src/utils.ts

```diff
@@ -183,7 +199,7 @@ export function areMapsEqual(
         b.forEach((bValue, bKey) => {
           if (!hasMatch && !matchedIndices[matchIndex]) {
             hasMatch =
-              isEqual(aKey, bKey, meta) && isEqual(aValue, bValue, meta);
+            isEqual(aKey, bKey, matchIndex, a, b, meta) && isEqual(aValue, bValue, aKey, a, b, meta);
```

Comment:
I just noticed `aKey` being used here instead of `matchIndex`, which creates a difference from how `Set` operates and also could be confusing since `b.get(aKey)` may return `false` even though the equality check could return `true`.

Reply:
You're right, it's not straightforward... This could return `undefined` if the key is not a primitive, and we're not using shallowEqual. It is inconsistent.
I think we get more information by passing `aKey` but the tradeoff is that inconsistency.
Having the same behaviour of passing the `matchIndex` for `Set` and `Map` would be confusing in my opinion, since Maps are supposed to be keyed.
A way to have it consistent between `Map` and `Set` would be to use arrays (converting) to compare them. This way it would be accessible with parent[index].

Reply:
## Option 1: always using `matchIndex`

An important distinction is that the definition of equality for this library differs from the spec for `Map` / `Set`. Example:

```ts
const mapA = new Map([[{ foo: 'bar' }, 'baz']]);
const mapB = new Map([[{ foo: 'bar' }, 'baz']]);

deepEqual(mapA, mapB); // true
```
This means that for `Map` specifically, providing only one or the other key would provide inaccurate information; `b.get(aKey)` with the above example would be invalid. This means confusion will always exist, just shifted dependency on the decision here.

Accepting that confusion will always exist, generally implementation consistency coupled with documentation is the chosen direction. With that in mind, something like this in the README section on `createCustomEqual`:

```md
**NOTE**: `Map` and `Set` implementations may have different key references and still be considered equal depending on your needs. Because order may be an aspect of the decision, the iteration index is provided as the `key` in the custom comparator. If you need to do a lookup in either object based on this index, it is recommended to use native methods like `.keys()` or `.values()` in combination with the index.
```

We can even provide an example, something like:

```js
function customComparator(aValue, bValue, iterationIndex, a, b, meta) {
  const aKey = Array.from(a.keys())[iterationIndex];
```

## Option 2: Embrace the difference in keys

We could optionally tweak the contract to support the different multiple keys:

```ts
export type InternalEqualityComparator = (
  objectA: any,
  objectB: any,
  indexOrKeyA: any,
  indexOrKeyB: any,
  parentA: any,
  parentB: any,
  meta: any,
) => boolean;
```

This would be duplicative in array / object operations:

```ts
// arrays
if (!isEqual(a[index], b[index], index, index, a, b, meta)) {
// objects
if (!hasOwnProperty(b, key) || !isEqual(a[key], b[key], key, key, a, b, meta)) {
```

But would allow full pass-through in `Map` / `Set` operations:

```ts
// maps
hasMatch = isEqual(aKey, bKey, matchIndex, matchIndex, a, b, meta) && isEqual(aValue, bValue, aKey, bKey, a, b, meta);
// sets
hasMatch = isEqual(aValue, bValue, aValue, bValue, a, b, meta);
```

At this point the only inconsistency would be the passing of `matchIndex` for doing `Map`-based key equality comparisons, however again documentation could explain this:

```md
**NOTE**: `Map` implementations compare equality for both keys and value. When using a custom comparator and comparing equality of the keys, the iteration index is provided as both `aKey` and `bKey` to help use-cases where ordering of keys matters to equality.
```

For all scenarios other than `Map` the multiple keys would not be much help, but the only cost would be a "throwaway" function parameter.

Reply:
> generally implementation consistency coupled with documentation is the chosen direction

I completely agree !

Using `matchIndex` has a problem : the index is only accurate for `b`. Sets and Maps don't take into account the order of elements when checking for equality and the index is incremented in the second loop, looping over b values. 
``` ts
  const setA = new Set(['x', 'y']);
  const setB = new Set(['y', 'x']);

  const createIsEqual: EqualityComparatorCreator =
    (deepEqual) => (a: any,
      b: any,
      indexOrKey?: any,
      parentA?: any,
      parentB?: any,
      meta?: any) => {
      expect(Array.from(parentB)[indexOrKey]).toBe(b);
      expect(Array.from(parentA)[indexOrKey]).toBe(a);

      return deepEqual(a, b, meta);
    };
  const comparator = createComparator(createIsEqual);
  comparator(setA, setB);
```
This test fails on the first `isEqual` call internally. The value passed is `'x'`, but at the index passed `1`, for `a`, the value is `'y'`

So I think we should go with option 2. Having 2 index/key params enables us to add another index that increments in the `a` loop and give two accurate indexes for doing Map-based key equality comparisons. What do you think ?

Reply:
Yup, agreed on all fronts!